PUBDEV-5336 - h2o python: get wrong ans when run a conditional statem…

Pscheidl opened the pull request:
…ent in h2o

Pscheidl commented on h2o-core/src/main/java/water/rapids/ast/prims/operators/AstBinOp.java:
```diff
@@ -329,15 +338,53 @@ public void map(Chunk[] chks, NewChunk[] cress) {
+   * @param targetDomain Domain to originally map to
+   * @return Cross-domain mapping as an array of primitive integers
+   */
+  private int[] alignCategoricals(String[] sourceDomain, String[] targetDomain) {
```
Having domains passed as method arguments ensures those are copies and original domain arrays are not affected while sorting.

Pscheidl commented on h2o-core/src/main/java/water/rapids/ast/prims/operators/AstBinOp.java:
```diff
@@ -407,7 +454,11 @@ public void map(Chunk[] chks, NewChunk[] cress) {
   }
 
-  // Make sense to run this OP on an enm?
+  /**
```
Took the opportuninty to add the comment as a JavaDoc.

Pscheidl commented on h2o-py/tests/testdir_jira/pyunit_pubdev_5336.py:
```diff
@@ -0,0 +1,33 @@
+    assert frame['eq'][2,0] == 0
+    assert frame['eq'][3,0] == 0
+
+    # Compare in inverse order (tests one more categorical variable in first column)
```
This indeed causes left and right chunk to be swapped, testing the correctness of mapping process (map from smaller to larger domain).

Pscheidl commented on h2o-core/src/main/java/water/rapids/ast/prims/operators/AstEq.java:
```diff
@@ -16,7 +16,8 @@
   }
-
+  
+  @Override
```
Took the opportunity to mark this method with `@Override` properly.

Pscheidl commented on h2o-core/src/main/java/water/rapids/ast/prims/operators/AstBinOp.java:
```diff
@@ -329,15 +338,53 @@ public void map(Chunk[] chks, NewChunk[] cress) {
+   */
+  private int[] alignCategoricals(String[] sourceDomain, String[] targetDomain) {
+    if (targetDomain.length > sourceDomain.length) {
+      String[] swap = targetDomain;
```
I'm not sure if swap goes out of scope on every JVM after this block is ended.

Pscheidl commented on h2o-core/src/main/java/water/rapids/ast/prims/operators/AstBinOp.java:
```diff
@@ -309,13 +310,25 @@ private ValFrame frame_op_frame(Frame lf, Frame rt) {
+
+    for (int c = 0; c < lf.numCols(); c++) {
+      if (lf.vec(c).isCategorical() && rt.vec(c).isCategorical()) {
+        if (lf.vec(c).domain().length >= rt.vec(c).domain().length) {
```
Pass arguments in right order beforehand to avoid swaping. For high count of domain levels, this may save some memory for a short period of time.

Pscheidl commented:
Failed tests working on Python 3.5. Strange. Adressing this issue.

Pscheidl commented on h2o-core/src/main/java/water/rapids/ast/prims/operators/AstBinOp.java:
```diff
@@ -329,15 +342,47 @@ public void map(Chunk[] chks, NewChunk[] cress) {
+          }
         }
       }
     }.doAll(lf.numCols(), Vec.T_NUM, new Frame(lf).add(rt)).outputFrame(lf._names, null);
```
Type should not be T_NUM for categorical ....

michalkurka commented on h2o-core/src/main/java/water/rapids/ast/prims/operators/AstBinOp.java:
```diff
@@ -329,15 +343,47 @@ public void map(Chunk[] chks, NewChunk[] cress) {
+   * @return Cross-domain mapping as an array of primitive integers
+   */
+  private int[] alignCategoricals(String[] longerDomain, String[] shorterDomain) {
+    Arrays.sort(longerDomain);
```
This is being called this way:

    alignCategoricals(lf.vec(c).domain(), rt.vec(c).domain())

Such call will mutate the domain of the Vec. This is undesirable because the original domain mapping will get lost. Vec will typically (99%) have domain already sorted but there is no guarantee (eg. a rename operation will produce an unordered domain).

Pscheidl replied:
The alignCategoricals method now creates a copy of the longerDomain before sorting it and performing binary search internally. It will never mutate any String[] argument - ensured. Sorting the shorter one is avoided of course !

michalkurka commented on h2o-core/src/main/java/water/rapids/ast/prims/operators/AstBinOp.java:
```diff
@@ -329,15 +343,47 @@ public void map(Chunk[] chks, NewChunk[] cress) {
+   */
+  private int[] alignCategoricals(String[] longerDomain, String[] shorterDomain) {
+    Arrays.sort(longerDomain);
+    Arrays.sort(shorterDomain);
```
There is no need to sort the `shorterDomain`. In fact one of the domain should not be sorted because you need to preserve the correct mapping!

Pscheidl replied:
Yes, the mutation is something I overlooked. Fixed ! Making a local copy of the longerDomain. Not sorting the other one as well (it worked out by accident, as it was already sorted before).

michalkurka commented on h2o-core/src/main/java/water/rapids/ast/prims/operators/AstBinOp.java:
```diff
@@ -329,15 +343,47 @@ public void map(Chunk[] chks, NewChunk[] cress) {
             for (int i = 0; i < clf._len; i++)
               cres.addNum(str_op(clf.atStr(lfstr, i), crt.atStr(rtstr, i)));
-          else
+          else if (categoricalOK() && clf.vec().isCategorical() && crt.vec().isCategorical()) {
```
This check doesn't have to be done for every row, I would just do it in the initialization. For readability purposes, I would also pass a boolean flag to the MRTask that will indicate whether to use left or right for the mapping. This way you can keep the decision login (`clf.vec().domain().length > crt.vec().domain().length`) at one place outside of this task.

Pscheidl replied:
These are done once before the MRTask is created (for categorical columns only). Then the values are red during iteration.

michalkurka commented on h2o-py/tests/testdir_jira/pyunit_pubdev_5336.py:
```diff
@@ -0,0 +1,33 @@
+    frame['Dest'].asfactor()
+
+    # First column has one more categorical variable
+    assert frame['Origin'].nlevels() == [3]
```
It would be nicer to check the exact domain, including the order. Eg. if H2O (erroneously) decides to order the domains [SFO, SAN, NYC] and [SFO, SAN] this PR would pass even without this change.

Pscheidl replied:
Added such tests.

michalkurka commented on h2o-core/src/main/java/water/rapids/ast/prims/operators/AstBinOp.java:
```diff
@@ -329,15 +343,47 @@ public void map(Chunk[] chks, NewChunk[] cress) {
+            // The vec with longer domain is iterated over due to categorical mapping
+            if (clf.vec().domain().length > crt.vec().domain().length) {
+              for (int i = 0; i < clf._len; i++) {
+                cres.addNum(op(clf.atd(i), alignedCategoricals[c][(int) crt.atd(i)]));
```
I think this will fail for NAs in the data. Casting Double.NaN to int will not work. NAs should be passed without transformation.

Pscheidl replied:
Indeed. Casting Double.NaN to 32-bit int results in the value of 0.

michalkurka commented on h2o-py/tests/testdir_jira/pyunit_pubdev_5336.py:
```diff
@@ -0,0 +1,33 @@
+
+
+def pubdev_5336():
+    data = pd.DataFrame({'Origin': ['SFO', 'SAN', 'SFO', 'NYC'],
```
The tests should also make sure NAs comparisons work.

Pscheidl commented:
In addition, the test will be enhanced with a brand new part that tests the original issue directly on airlines_train.csv, where the number of equal origins and destinations should be zero.

Pscheidl commented:
The alignment of categorical variables has been moved after the sanity checks. No need to align huge domains only to throw the results away few lines later.

Pscheidl commented:
NaNs are now detected. NaNs are just passed to the op(operand1, operand2) function and not treated as indices to categorical value mapping array. This makes it possible for two NaNs to actually be marked as equal.

michalkurka commented on h2o-core/src/main/java/water/rapids/ast/prims/operators/AstBinOp.java:
```diff
@@ -329,15 +350,58 @@ public void map(Chunk[] chks, NewChunk[] cress) {
+            if (rtDomainNotBigger[c]) {
+              for (int i = 0; i < clf._len; i++) {
+                double crtAtdValue = crt.atd(i);
+                if (Double.isNaN(crtAtdValue)) {
```
This should be `crt.isNA(i)`, relying on NaN is an API misuse.

michalkurka commented on h2o-py/tests/testdir_jira/pyunit_pubdev_5336.py:
```diff
@@ -0,0 +1,42 @@
+
+    # First column has one more categorical variable
+    assert frame['Origin'].nlevels() == [4]
+    assert frame['Origin'].levels() == [['NYC', 'SAN', 'SFO', 'nan']];
```
This shows the test setup is actually incorrect - NA cannot be part of the domain.

michalkurka commented:
@Pscheidl you need to handle NAs (not NaNs). I hope I didn't confuse you in the previous iteration of the review where I mentioned NaNs. NaN is used internally (but not always!) to represent an NA. You need to use the Chunk API to find out if a value is NA or not.

michalkurka commented:
Note to self: Not updated yet

Pscheidl commented:
@michalkurka 
Thanks for your assistance. Corrected the NA checks, removed the API abuse and finally added a test with actual NAs. Reused currently present prostate_missing.csv.

Testing both comparison on lines with two NAs on columns forced to be categorical (should be true) and only one-sided comparison. The NA variables are actually recognized by the crt.isNA(i) (or clf.isNA(i)) check.

Please review and let's wait for tests.
